### PR TITLE
Get proper index for padded broadcast dimension

### LINKED ISSUE #868: Pad of Broadcast IterDomain broadcasts into the pad region
PR #841 enabled us to `cat` along broadcast `IterDomain`s, by allowing those pads. In that case, the pad values are unused since they are predicated away. If we actually inspect the pad values, it seems they are incorrect: the broadcast value is being used in place of the pad value. See below:

```python
import torch
import torch.nn.functional as F
from nvfuser import FusionDefinition, DataType

torch.manual_seed(0)
inputs = [
        torch.randn((1,), device='cuda'),
        ]

with FusionDefinition() as fd:
    x = fd.from_pytorch(inputs[0])
    x_pad = fd.ops.pad(x, [1, 1])
    fd.add_output(x_pad)

nvf_out = fd.execute(inputs)
# With NVFUSER_DUMP=fusion_ir_preseg:
# Inputs:
#   T0_g[ bS0{1} ], float
# Outputs:
#   T1_g[ iS2{( ( ( (nvfuser_index_t)(1) ) + 1 ) + ( (nvfuser_index_t)(1) ) )}rf ], float
# 
# %kernel_math {
# i6 = (nvfuser_index_t)(1);
# i8 = (nvfuser_index_t)(1);
# T1_g[ iS2{( ( ( (nvfuser_index_t)(1) ) + 1 ) + ( (nvfuser_index_t)(1) ) )}rf ]
#    = pad( T0_g[ bS0{1} ], {i6, i8} )
# }

print(nvf_out[0])
# tensor([-0.9247, -0.9247, -0.9247], device='cuda:0')
# Should be tensor([0, -0.9247, 0], device='cuda:0')

assert(torch.equal(nvf_out[0], F.pad(inputs[0], (1, 1))))
```
We see that the output `IterType` is `Iteration`, which rules out the possibility that the output was interpreted as an expanded broadcast instead of a pad.

Related to #21

### PR DESCRIPTION
**Pull request by jacobhinkle**
This PR:

- Changes `IndexLowering::handle(PadOp*)` to override the index of padded `Broadcast` dimensions and use the usual `Iteration` index instead of `zeroVal()`.
- Changes `Index::getProducerAllocationIndices` to only skip broadcast dimensions if not overridden, so that the above change to `IndexLowering` has an effect.
- Updates frontend `test_pad` test to exercise pad of a broadcasted dim and adds C++ tests padding broadcasts and expanded broadcasts.

Fixes #868.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build   ajdklsdf

File: csrc/device_lower/pass/index.cpp [hunk 1]
```diff
@@ -1471,8 +1473,54 @@ void IndexLowering::handle(const PadOp* pad) {
...
+      if (index_it != index_map.end()) {
+        std::cout << "Overriding index for " << padded_id->toString() << " as "
+                  << index_it->second->toInlineString() << std::endl;
+        override_index.emplace(padded_id, index_it->second);
+      } else {
+        std::cout << "Couldn't find index for " << padded_id->toString()
+                  << std::endl;
+      }
+    }
+  }
```
Comment (jacobhinkle) [inline csrc/device_lower/pass/index.cpp:1520]: This is redundant and brittle. An alternative would be to add an `const std::vector<size_t>& ignore_broadcasts = {}` arg to `Index::getProducerPerDimLogicalIndex`. I did it this way at first since there is already `override_index`.

File: csrc/device_lower/analysis/index_compute.cpp [hunk 2]
```diff
@@ -838,7 +838,7 @@ IndexFromIdGraph getTensorIndexFromIdGraph(
     const TensorView* consumer_tv,
     const TensorView* producer_tv,
     bool is_global,
-    std::unordered_map<IterDomain*, IterDomain*> c2p_map) {
+    const std::unordered_map<IterDomain*, IterDomain*> c2p_map) {
```
Comment (jacobhinkle) [inline csrc/device_lower/analysis/index_compute.cpp:841]: Unrelated to this PR, but clang-tidy asked me to fix some const-correctness, so I fixed some nearby missing const while I was at it.

**Comment by jacobhinkle** [comment]
!build

File: csrc/device_lower/analysis/index_compute.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/device_lower/analysis/index_compute.cpp:841]: Should this be `const std::unordered_map<IterDomain*, IterDomain*>&` instead of `const std::unordered_map<IterDomain*, IterDomain*>`?

File: csrc/device_lower/analysis/index_compute.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ IndexFromIdGraph getTensorIndexFromIdGraph(
     const TensorView* consumer_tv,
     const TensorView* producer_tv = nullptr,
     bool is_global = true,
-    std::unordered_map<IterDomain*, IterDomain*> c2p_map = {});
+    const std::unordered_map<IterDomain*, IterDomain*> c2p_map = {});
```
Comment (zasdfgbnm) [inline csrc/device_lower/analysis/index_compute.h:44]: same as https://github.com/NVIDIA/Fuser/pull/869/files#r1337566867

File: csrc/device_lower/analysis/index_compute.cpp (hunk 2 above)
Comment (jacobhinkle) [reply csrc/device_lower/analysis/index_compute.cpp:841]: Nice catch. I wonder why `clang-tidy` doesn't find that??

File: csrc/device_lower/pass/index.cpp [hunk 4]
```diff
@@ -1479,8 +1520,21 @@ void IndexLowering::handle(const PadOp* pad) {
...
   const auto pad_val = pad->value();
 
+  std::unordered_map<IterDomain*, Val*> override_index;
+  for (auto padded_axis : pad->getPaddedAxes()) {
+    auto padded_id = producer_doms.at(padded_axis);
+    if (padded_id->isBroadcast()) {
+      // When we pad a Broadcast IterDomain, we should not treat it as a
+      // Broadcast as we normally would. Instead, we will treat it as a regular
+      // Iteration domain with extent 1.
+      auto ind =
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/index.cpp:1530]: Instead of copy-pasting a function, does it make sense to make a clone of the tensor domain, and use 
https://github.com/NVIDIA/Fuser/blob/fb0e25e50f427305972de9d96dda21fa3606c5d7/csrc/device_lower/utils.h#L47
to temporarily modify the tensor domain?

Comment (jacobhinkle) [reply csrc/device_lower/pass/index.cpp:1530]: Oh I hadn't even considered doing that. Let me try it out!

Comment (jacobhinkle) [reply csrc/device_lower/pass/index.cpp:1530]: This method is slightly complicated since we need to create a replacement `IterDomain` for each broadcast `IterDomain` in the producer's rfactor domain that is being padded, and that replacement must be derived from the root domain or else we get an error when trying to create a replacement `TensorDomain`. That means we then need to actually replace all the siblings of each padded broadcast ID, and so on.

File: csrc/device_lower/pass/index.cpp [hunk 5]
```diff
@@ -1479,8 +1520,21 @@ void IndexLowering::handle(const PadOp* pad) {
 
   const auto pad_val = pad->value();
 
+  std::unordered_map<IterDomain*, Val*> override_index;
+  for (auto padded_axis : pad->getPaddedAxes()) {
+    auto padded_id = producer_doms.at(padded_axis);
+    if (padded_id->isBroadcast()) {
+      // When we pad a Broadcast IterDomain, we should not treat it as a
```
Comment (naoyam) [inline csrc/device_lower/pass/index.cpp:1527]: Trying to understand why. Is this only necessary for the padding predicate? For the normal indexing, we shouldn't need to index into the broadcast domain, right?

Comment (jacobhinkle) [reply csrc/device_lower/pass/index.cpp:1527]: Yes it's just for the padding predicate. Otherwise we will have a constant 0 which means the padding predicate is always true, so we never use the pad value.

Comment (naoyam) [reply csrc/device_lower/pass/index.cpp:1527]: I wonder if we could minimize the usage of `override_index`. That's a workaround for ops like `index_select`, and it shouldn't be used unless absolutely necessary.

Since it should result in `0`, we may stop kipping broadcast domains. I believe that should give us a correct predicate for padding, and we still end up having the same predicates and indices in anywhere else.

Comment (naoyam) [reply csrc/device_lower/pass/index.cpp:1527]: One correction. It's already using `override_index` only for generating the pad predicate.

Comment (naoyam) [reply csrc/device_lower/pass/index.cpp:1527]: Sorry, my comment above was completely wrong. Still trying to get a better understanding...

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the fix. While Having more uses of the index overriding is concerning, I don't see any other way.
